File: app/src/main/java/com/stathis/elmepaunivapp/ui/announcements/AnnouncementsViewModel.java
package com.stathis.elmepaunivapp.ui.announcements;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import androidx.lifecycle.ViewModel;

import com.stathis.elmepaunivapp.database.AnnouncementsDatabase;
import com.stathis.elmepaunivapp.listeners.activity_listeners.AnnouncementClickListener;
import com.stathis.elmepaunivapp.listeners.NewsClickListener;
import com.stathis.elmepaunivapp.ui.announcements.model.Announcement;
import com.stathis.elmepaunivapp.ui.announcements.recyclerviews.LatestNewsAdapter;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class AnnouncementsViewModel extends ViewModel implements NewsClickListener {

    AnnouncementsDatabase announcementsDatabase;
    LatestNewsAdapter adapter = new LatestNewsAdapter(this);
    ArrayList<Announcement> announcements = new ArrayList<>();
    Content content;
    private AnnouncementClickListener announcementClickListener;

    void setUpListener(AnnouncementClickListener announcementClickListener) {
        this.announcementClickListener = announcementClickListener;
    }

    void setupDatabase(Context context) {
        announcementsDatabase = AnnouncementsDatabase.getInstance(context);
    }

    void getAnnouncements() {
        adapter.submitList(announcementsDatabase.getAnnouncementDao().getAll());

        content = new Content();
        content.execute();
    }

    @Override
    public void onNewsClick(Announcement announcement) {
        announcementClickListener.goToAnnouncement(announcement);
    }

    public boolean dbIsEmpty() {
        List<Announcement> announcements = announcementsDatabase.getAnnouncementDao().getAll();
        if (announcements.size() < 1) {
            return true;
        }
        return false;
    }

    private class Content extends AsyncTask<Void, Void, Void> {

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);

            adapter.submitList(announcementsDatabase.getAnnouncementDao().getAll());
            adapter.notifyDataSetChanged();
        }

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                String url = "https://mst.hmu.gr/news_gr/";
                Document doc = Jsoup.connect(url).get();
                Elements data = doc.select("article");
                int size = data.size();
                Log.d("doc", "doc: " + doc);
                Log.d("data", "data: " + data);
                for (int i = 0; i < 13; i++) {
                    String imgUrl = data.select("a.entry-featured-image-url")
                            .select("img")
                            .eq(i)
                            .attr("src");

                    String title = data.select("h2.entry-title")
                            .select("h2")
                            .eq(i)
                            .text();

                    String detailUrl = data.select("h2.entry-title")
                            .select("a")
                            .eq(i)
                            .attr("href");
                    announcements.add(new Announcement(title, detailUrl, imgUrl));
                    Log.d("items", "img: " + imgUrl + " . title: " + title);
                }

                /* checks if announcements db size is less than 0 (empty) and either inserts or updates the data
                then passes the data into the ListAdapter   */

                if (dbIsEmpty()) {
                    announcementsDatabase.getAnnouncementDao().insert(announcements);
                } else {
                    announcementsDatabase.getAnnouncementDao().update(announcements);
                }

            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }
    }
}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/announcements/AnnouncementsViewModel.kt
package com.stathis.elmepaunivapp.ui.announcements

import android.app.Application
import android.util.Log
import android.view.View
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.ElmepaViewModel
import com.stathis.elmepaunivapp.callbacks.AnnouncementClickListener
import com.stathis.elmepaunivapp.callbacks.ElmepaClickListener
import com.stathis.elmepaunivapp.database.AnnouncementsDatabaseKt
import com.stathis.elmepaunivapp.model.Announcement
import com.stathis.elmepaunivapp.ui.announcements.adapter.AnnouncementsAdapter
import com.stathis.elmepaunivapp.util.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.jsoup.Jsoup

class AnnouncementsViewModel(app: Application) : ElmepaViewModel(app), ElmepaClickListener {

    private lateinit var callback: AnnouncementClickListener
    val adapter = AnnouncementsAdapter(this)
    val announcements = MutableLiveData<List<Announcement>>()
    val error = MutableLiveData<Boolean>()
    val database = AnnouncementsDatabaseKt.getDatabase(app).announcementDao()
    private val prefHelper = SharedPreferencesHelper.setHelper(app)
    private val refreshTime = 5 * 60 * 1000 * 1000 * 1000L

    /*
     *  Data fetch logic:
     *  Trying to get data from db,  if this fails, then get from web.
     *  The app decides when to fetch data from web or db from the timer implemented
     */

    init {
        startShimmer()
    }

    private fun startShimmer(){
        adapter.submitList(ShimmerHelper.list)
    }

    fun bindCallback(callback: AnnouncementClickListener) {
        this.callback = callback
    }

    fun refreshData() {
        getData()
    }

    fun observe(owner: LifecycleOwner) {
        announcements.observe(owner) {
            it?.let { adapter.submitList(it) }
        }
    }

    fun release(owner: LifecycleOwner) {
        announcements.removeObservers(owner)
    }

    private fun getData() {
        viewModelScope.launch(Dispatchers.IO) {
            val updateTime = SharedPreferencesHelper.getUpdateTime()
            val currentTime = System.nanoTime()

            when (updateTime > 0 && currentTime - updateTime < refreshTime) {
                true -> getAnnouncementsFromDb() //Getting Data From Database
                false -> getAnnouncements() //Getting Data From Web
            }
        }
    }

    private fun getAnnouncementsFromDb() {
        val announcementList = database.getAll()
        announcements.postValue(announcementList)
    }

    private fun getAnnouncements() {
        val announcementList = arrayListOf<Announcement>()

        try {
            val doc = Jsoup.connect(BASE_URL).get().select(DATA_TYPE)
            for (i in 0..12) {
                val imgUrl = doc.select(IMG_HTML_TAG).select(IMG_TYPE).eq(i).attr(IMG_SOURCE)
                val title = doc.select(TITLE_HTML_TAG).select(TITLE_TYPE).eq(i).text()
                val detailUrl = doc.select(URL_HTML_TAG).select(URL_TYPE).eq(i).attr(URL_ATTR)
                announcementList.add(Announcement(title, detailUrl, imgUrl))
            }

            announcements.postValue(announcementList)

            database.deleteAll()
            database.insertAll(announcementList)

            SharedPreferencesHelper.saveUpdateTime(System.nanoTime())
        } catch (e: Exception) {
            Log.d(getString(R.string.app_name), e.toString())
            getAnnouncementsFromDb()
        }
    }

    override fun onItemClick(view: View) {
        when (view.tag) {
            is Announcement -> callback.onAnnouncementTap(view.tag as Announcement)
        }
    }
}
